components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,322 @@
+    //                        }
+    //                    }
+    //                }
+    //                return Ok(false);
+    //            }
+    //            None => Ok(false),
+    //        },
+    //        Err(e) => {
+    //            debug!("{:?}", e);
+    //            Err(Error::BadBucket(e))
+    //        }
+    //    }
+    //}
+
+    // This function checks whether or not the
+    // configured bucket exists in the configured
+    // backend.
+    fn exists(&self) -> Result<bool> {
+        let artifactbucket = self.bucket.to_owned();
+        match self.client.list_buckets().sync() {
+            Ok(bucket_list) => match bucket_list.buckets {
+                Some(buckets) => {
+                    for bucket in buckets {
+                        if bucket.name.unwrap() == artifactbucket {
+                            return Ok(true);
+                        } else {
+                            return Ok(false);
+                        }
+                    }
+                    return Ok(false);
+                }
+                None => Ok(false),
+            },
+            Err(e) => {
+                debug!("{:?}", e);
+                return Err(Error::ListBuckets(e));
+            }
+        }
+    }
+
+    pub fn create_bucket(&self) -> Result<()> {
+        let mut request = CreateBucketRequest::default();
+        request.bucket = self.bucket.clone();
+
+        if self.exists().unwrap() {
+            Ok(())
+        } else {
+            match self.client.create_bucket(&request).sync() {
+                Ok(_response) => Ok(()),
+                Err(e) => {
+                    debug!("{:?}", e);
+                    Err(Error::CreateBucketError(e))
+                }
+            }
+        }
+    }
+
+    pub fn upload(
+        &self,
+        hart: &PathBuf,
+        ident: &PackageIdent,
+        target: &PackageTarget,
+    ) -> Result<()> {
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        let file = File::open(hart)?;
+        let size = file.metadata().unwrap().len() as usize;
+        let min_limit: usize = 49152 * 1024;
+        if size < min_limit {
+            self.single_upload(key, file)
+        } else {
+            self.multipart_upload(key, file)
+        }
+    }
+
+    pub fn download(
+        &self,
+        loc: &PathBuf,
+        ident: &PackageIdent,
+        target: &PackageTarget,
+    ) -> Result<()> {
+        let mut request = GetObjectRequest::default();
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        request.bucket = self.bucket.to_owned();
+        request.key = key;
+
+        let payload = self.client.get_object(&request).sync();
+        let body = match payload {
+            Ok(response) => response.body,
+            Err(e) => {
+                return Err(Error::PackageDownload(e));
+            }
+        };
+
+        let file = body.expect("Downloaded pkg archive empty!").concat2();
```

Comment:
Check result instead of expect..?

Reply:
This isn't actually a result, the result was pulled out before this expect, this is actually an expectation of an Option. Is using a match here still what you'd like me to do?

Reply:
Ah ok, seems safe as we're never pushing empty packages.. just don't know what to expect with S3 APIs :)